File: frontend/chat/src/turnPresentation.ts
import type { ChatMessage } from "./types";

export type MemoryHit = { id: string; summary: string };
export type MemoryResult = { count: number; items: MemoryHit[] };
export type ProcessPart =
  | { id: string; kind: "thinking" | "text" | "answer"; text: string }
  | { id: string; kind: "tool"; call_id: string; memory_result?: MemoryResult }
  | { id: string; kind: "memory"; query: string; items: MemoryHit[]; duration_ms?: number };
export type TurnPresentation = { version: 1; started_at?: string; final?: boolean; parts: ProcessPart[] };

/** 展示元数据向后兼容；未知版本不影响原有正文与工具记录。 */
export function readPresentation(value: unknown): TurnPresentation | undefined {
  if (!value || typeof value !== "object") return undefined;
  const data = value as Record<string, unknown>;
  if (data.version !== 1 || !Array.isArray(data.parts)) return undefined;
  const parts: ProcessPart[] = [];
  const ids = new Set<string>();
  for (const raw of data.parts) {
    if (!raw || typeof raw !== "object") continue;
    const part = raw as Record<string, unknown>;
    if (typeof part.id !== "string" || ids.has(part.id)) continue;
    if ((part.kind === "thinking" || part.kind === "text" || part.kind === "answer") && typeof part.text === "string") {
      parts.push({ id: part.id, kind: part.kind, text: part.text });
    } else if (part.kind === "tool" && typeof part.call_id === "string") {
      const memory = readMemoryResult(part.memory_result);
      parts.push({ id: part.id, kind: "tool", call_id: part.call_id, ...(memory ? { memory_result: memory } : {}) });
    } else if (part.kind === "memory" && Array.isArray(part.items)) {
      const items = readMemoryHits(part.items);
      if (items.length) parts.push({ id: part.id, kind: "memory", query: String(part.query ?? ""), items,
        ...(typeof part.duration_ms === "number" && part.duration_ms >= 0 ? { duration_ms: part.duration_ms } : {}) });
    }
    ids.add(part.id);
  }
  return { version: 1, parts, final: data.final === true,
    ...(typeof data.started_at === "string" ? { started_at: data.started_at } : {}) };
}

export function readMemoryHits(items: unknown[]): MemoryHit[] {
  return items.flatMap((item) => {
    if (!item || typeof item !== "object") return [];
    const record = item as Record<string, unknown>;
    return typeof record.summary === "string" ? [{ id: String(record.id ?? ""), summary: record.summary }] : [];
  });
}

export function readMemoryResult(value: unknown): MemoryResult | null {
  if (!value || typeof value !== "object") return null;
  const result = value as Record<string, unknown>;
  if (!Array.isArray(result.items)) return null;
  const items = readMemoryHits(result.items);
  return { count: typeof result.count === "number" && result.count >= 0 ? result.count : items.length, items };
}

export function appendProcessDelta(presentation: TurnPresentation | undefined, id: string | undefined,
  kind: "thinking" | "text", delta: string): TurnPresentation | undefined {
  if (!id) return presentation;
  const base = presentation ?? { version: 1 as const, parts: [] };
  const index = base.parts.findIndex((part) => part.id === id);
  const parts = [...base.parts];
  const previous = parts[index];
  if (previous && previous.kind === kind) parts[index] = { ...previous, text: previous.text + delta };
  else if (index < 0) parts.push({ id, kind, text: delta });
  return { ...base, parts };
}

export function messageProcess(message: ChatMessage): { parts: ProcessPart[]; body: string } {
  const presentation = message.presentation;
  if (!presentation) {
    // 旧消息没有逐段时序，只保留原来的思考与工具顺序，不推测思考的分段位置。
    return { parts: [
      ...(message.thinking ? [{ id: "legacy-thinking", kind: "thinking" as const, text: message.thinking }] : []),
      ...message.tools.map((tool) => ({ id: `tool-${tool.callId}`, kind: "tool" as const, call_id: tool.callId })),
    ], body: message.content };
  }
  const parts = presentation.parts.filter((part) => part.kind !== "answer");
  if (presentation.final) return { parts, body: message.streaming
    ? presentation.parts.flatMap((part) => part.kind === "answer" ? [part.text] : []).join("")
    : message.content };
  // 尚未获得结束原因时，最后一段文字照常实时显示；后续工具到达后自然成为过程文字。
  const last = parts.at(-1);
  if ((message.streaming || message.status === "interrupted") && last?.kind === "text") {
    return { parts: parts.slice(0, -1), body: last.text };
  }
  return { parts, body: message.streaming || message.status === "interrupted" ? "" : message.content };
}

File: agent/turn_presentation.py
"""独立于模型上下文的会话展示投影，只记录真实发生的过程。"""

from __future__ import annotations

from copy import deepcopy
import json
from typing import Any

from agent.tool_projection import project_result_preview


class TurnPresentation:
    def __init__(self, started_at: str) -> None:
        self.data: dict[str, Any] = {"version": 1, "started_at": started_at, "parts": []}

    def append_text(self, iteration: int, kind: str, text: str) -> str:
        parts = self.data["parts"]
        if parts and parts[-1]["kind"] == kind and parts[-1].get("iteration") == iteration:
            parts[-1]["text"] += text
            return parts[-1]["id"]
        part_id = f"part-{len(parts)}"
        parts.append({"id": part_id, "kind": kind, "iteration": iteration, "text": text})
        return part_id

    def resolve_response(self, iteration: int, content: str, thinking: str, *, final: bool) -> None:
        # 非流式响应和工具声明后被供应商抑制的正文在完成边界补齐，不重放 delta。
        for kind, text in (("thinking", thinking), ("text", content)):
            matching = [p for p in self.data["parts"] if p.get("iteration") == iteration and p["kind"] == kind]
            if text and not matching:
                self.append_text(iteration, kind, text)
            elif matching and "".join(p["text"] for p in matching) != text:
                matching[0]["text"] = text
                for part in matching[1:]:
                    part["text"] = ""
        if final:
            self.data["final"] = True
            for part in self.data["parts"]:
                if part.get("iteration") == iteration and part["kind"] == "text":
                    part["kind"] = "answer"

    def add_tool(self, call_id: str) -> None:
        self.data["parts"].append({"id": f"tool-{call_id}", "kind": "tool", "call_id": call_id})

    def complete_memory_tool(self, call_id: str, result: str) -> bool:
        """在通用结果截断前提取摘要，保持原工具输出和存储协议不变。"""
        try:
            value = json.loads(result)
        except (ValueError, TypeError):
            return False
        if not isinstance(value, dict) or not isinstance(value.get("items"), list):
            return False
        items = _memory_items(value["items"])
        for part in self.data["parts"]:
            if part.get("call_id") == call_id:
                part["memory_result"] = {"count": len(value["items"]), "items": items}
                return True
        return False

    def add_memory(self, details: dict[str, Any], duration_ms: int) -> None:
        if details.get("items"):
            self.data["parts"].append({
                "id": "automatic-memory", "kind": "memory", "query": project_result_preview(details["query"]),
                "items": _memory_items(details["items"]), "duration_ms": duration_ms,
            })

    def snapshot(self) -> dict[str, Any]:
        return deepcopy(self.data)


def _memory_items(items: list[Any]) -> list[dict[str, str]]:
    return [
        {"id": str(item.get("id") or ""), "summary": project_result_preview(item.get("summary"), max_length=1500)}
        for item in items if isinstance(item, dict) and isinstance(item.get("summary"), str)
    ]
